Adds `DeepNetFeatures` (#460)
plenoptic-org/plenoptic: docs/user_guide/models_and_metrics/feature_extractor.md

---
jupytext:
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.17.3
kernelspec:
  display_name: plenoptic
  language: python
  name: python3
---

```{code-cell} ipython3
:tags: [hide-input]

import pooch

# don't have pooch output messages about downloading or untarring
logger = pooch.get_logger()
logger.setLevel("WARNING")
```

:::{admonition} Run this notebook yourself!
:class: important

Download the script: [`feature_extractor.py`](../../scripts/feature_extractor.py)!

:::

(feature_extractor)=

> # Using Deep Neural Networks with plenoptic

Reviewer: “Using deep neural networks with plenoptic” is quite general for a title. I feel like we can almost call it “Reproducing Feather et al., 2023”. (again, depends on whether we want to treat this as a reproduction example, see my general comment in the PR thread)

Reviewer 2: My two cents: I like the current title. The tutorial shows how to use DNNs with plenoptic, which is a very important use case, so this should be easily findable by the title.

:::{warning}
This notebook requires the optional dependency `torchvision`, which can be installed with `pip`.
:::

> plenoptic is compatible with any model written in pytorch, including deep neural networks from the model zoos {external+torchvision:ref}`TorchVision <models>` and {external+timm:doc}`timm <models>`. In this notebook, we'll show how to adapt a deep net from these two packages for use with plenoptic, recreating some ResNet50 metamers shown in {cite:alp}`Feather2023-model-metam`, figure 2e.

Reviewer: ```suggestion
Plenoptic is compatible with any model written in pytorch, including deep neural networks from the model zoos {external+torchvision:ref}`TorchVision <models>` and {external+timm:doc}`timm <models>`. In this notebook, we'll show how to adapt a deep net from these two packages for use with plenoptic, recreating some ResNet50 metamers shown in {cite:alp}`Feather2023-model-metam`, figure 2e.
```

> plenoptic is compatible with any model written in pytorch, including deep neural networks from the model zoos {external+torchvision:ref}`TorchVision <models>` and {external+timm:doc}`timm <models>`. In this notebook, we'll show how to adapt a deep net from these two packages for use with plenoptic, recreating some ResNet50 metamers shown in {cite:alp}`Feather2023-model-metam`, figure 2e.

Reviewer 2: ```suggestion
plenoptic is compatible with any model written in pytorch, including deep neural networks. In this notebook we show how to use `plenoptic` with models from the deep network zoos {external+torchvision:ref}`TorchVision <models>` and {external+timm:doc}`timm <models>` to recreate some ResNet50 metamers shown in {cite:alp}`Feather2023-model-metam`, figure 2e.
```

```{code-cell} ipython3
import matplotlib.pyplot as plt
import torch

import plenoptic as po

# this notebook uses torchvision, which is an optional dependency.
# if this fails, install torchvision in your plenoptic environment
# and restart the notebook kernel.
try:
    import torchvision
except ModuleNotFoundError:
    raise ModuleNotFoundError(
        "optional dependency torchvision not found!"
        " please install it in your plenoptic environment "
        "and restart the notebook kernel"
    )

dtype = torch.float32
DEVICE = torch.device("cuda" if torch.cuda.is_available() else "cpu")

%load_ext autoreload

%autoreload 2

# so that relative sizes of axes created by po.plot.imshow and others look right
plt.rcParams["figure.dpi"] = 72

# set seed for reproducibility
po.set_seed(0)
```

:::{admonition} This notebook retrieves cached synthesis results
:class: warning dropdown

The example metamer shown in this notebook takes about 15 minutes to synthesize on a GPU. Thus, instead of performing synthesis in this notebook, we have cached the result of it online and only download them for investigation.

> :::

Reviewer 2: ```suggestion
:::{admonition} Executing this notebook takes a long time
:class: important

The example metamer shown in this notebook takes about 15 minutes to synthesize on a GPU.
:::
```

Reviewer 2: I think that the cache detail is too much detail for the user, although it is useful to mention that the notebook takes a while to run.

Author: I disagree -- I refer to the cache this and other notebooks (Portilla-Simoncelli, Berardino replication) to be clear why we're not calling synthesize.

I've now split up the notebook into two sections, one for "how to use deep nets" and one for "replicating Feather et al, 2023", and the cache warning is only on the Feather notebook (since the "how to use deep nets" one does call synthesize). I think this works, let me know what you think

Reviewer 2: I didn't realize we weren't calling synthesize. On my quick read, I thought you were adding the code for the synthesize call, but were not actually running it in the docs generation, and that that's what you were explaining to the user.

I think that, ideally, the notebook should do that: Have a sham synthesize call, but in the background actually load cached results, and not mention it. But even if you would consider this idea, I think it should be done in a future PR.

But for this PR, shouldn't this admonition be in the **Initialize and load Metamer objects** section, which is where the relevant code appears?

Reviewer 2: I didn't realize we weren't calling synthesize, I though you were just doing the cache loading in the background. My bad.

I agree then with keeping the original admonition. Ideally, though, I think that the notebook should just have the code to do the synthesis. It could be a sham synthesis call, with the cache loading happening in the background, or have these docs synthesize the metamer but not be built on every push. But even if you would consider this idea, it should be done in a future PR.

But agreeing to keep the admonition for this PR, shouldn't it be in the **Initialize and load Metamer objects** section, which is where the relevant code appears?

> ## Initializing the model

Reviewer 2: Add subsection headers for "Initialize model weights", "Select layer" and "Specify preprocessing" and "Prepare image".

I think this would make the tutorial A LOT more readable.

> When synthesizing images for deep nets, as in {cite:alp}`Feather2023-model-metam`, it is common to pick a specific intermediate layer whose representation we wish to use. `torchvision` contains a "feature extractor" to grab activity from intermediate layers, and plenoptic's {class}`plenoptic.models.FeatureExtractorModel` is a small wrapper to simplify this process.

Reviewer: Would be nice to link “feature extractor” to the torch documentation

> When synthesizing images for deep nets, as in {cite:alp}`Feather2023-model-metam`, it is common to pick a specific intermediate layer whose representation we wish to use. `torchvision` contains a "feature extractor" to grab activity from intermediate layers, and plenoptic's {class}`plenoptic.models.FeatureExtractorModel` is a small wrapper to simplify this process.

Reviewer 2: ```suggestion
To synthesize images for deep nets, we have to pick a specific layer whose representation we wish to use. {class}`plenoptic.models.FeatureExtractorModel` allows to easily grab activity from a deep net intermediate layers.
```

Reviewer 2: More focused emphasis on what the model subject of this tutorial does, rather than the secondary info of it being a wrapper around another's library similar functionality.

::::{tab-set}
:::{tab-item} layer2
:sync: layer2

```python
target_layer = "layer2"
```
:::
:::{tab-item} layer3
:sync: layer3

```python
target_layer = "layer3"
```
:::
:::{tab-item} layer4
:sync: layer4

```python
target_layer = "layer4"
```
:::

> ::::

Reviewer 2: These tabs that select the layer seem out of place here. Move to after the paragraph starting with "Next, we need to specify the layer to target." (current line 146 https://github.com/plenoptic-org/plenoptic/pull/460/changes#diff-290f4a0fc1cbf8ae472da0fdeefd5a319069b23365c324d20d3544aa76da96dcR146)

> In the rest of this section, we show how to initialize a plenoptic-compatible model using the weights from either the {external+torchvision:ref}`TorchVision <models>` or {external+timm:doc}`timm <models>` model zoos; their behavior after this section is the same.

Reviewer 2: ```suggestion
In the rest of this section, we show how to initialize a plenoptic-compatible model using the weights from either the {external+torchvision:ref}`TorchVision <models>` or {external+timm:doc}`timm <models>` model zoos. For that we need to specify three things: 1) The deep net model, 2) the layer(s) to extract, and 3) the image pre-processing to use.
```

First, we download the model weights for ResNet50 trained on [ImageNet-1K](https://en.wikipedia.org/wiki/ImageNet#ImageNet-1K) and initialize the `torchvision` / `timm` model.

::::{tab-set}
:::{tab-item} torchvision
:sync: torchvision

```{literalinclude} ../../scripts/feature_extractor.py
:dedent:
:start-at: ResNet50_Weights
:end-at: weights=weights
```

:::

:::{tab-item} timm
:sync: timm

Note that to run this cell (and the following `timm` cells), you must install `timm` as well (`pip install timm`)!

```{literalinclude} ../../scripts/feature_extractor.py
:dedent:
:start-at: timm.create_model
:end-at: timm.create_model
```
:::
::::

Next, we ensure that our model is in evaluation mode. Many models, including ResNet50, behave differently when in training and evaluation mode. In plenoptic, models are fixed and so we want the evaluation behavior:

```{literalinclude} ../../scripts/feature_extractor.py
:dedent:
:start-at: deepnet.eval()
:end-at: deepnet.eval()
```

> Next, we need to specify the layer to target. If we look at the ResNet50 metamers in Figure 2e from {cite:alp}`Feather2023-model-metam`, we can see an interesting progression in layers 2 through 4: the layer 2 metamer looks almost identical to the target image, the layer 3 metamer starts to add RGB noise, and the layer 4 is almost completely unidentifiable, looking almost completely like random RGB noise.

Reviewer 2: ```suggestion
Next, we specify the layer to target. Figure 2e from {cite:alp}`Feather2023-model-metam` shows an interesting progression for ResNet50 metamers: the layer 2 metamer looks almost like the target image, the layer 3 metamer shows some RGB noise, and the layer 4 metamer is almost completely unidentifiable.
```

> Let's start with `"layer3"`, but note the metamer synthesis procedure in this notebook works with any of `"layer2"`, `"layer3"`, or `"layer4"` (and possibly others, they just haven't been tested).

Reviewer 2: ```suggestion
Let's start with `"layer3"`, but note that the metamer synthesis procedure in this notebook works with any of `"layer2"`, `"layer3"`, or `"layer4"`, and that you can also specify multiple layers simultaneously.
```

:::{admonition} How do I know what layers I can use?
:class: dropdown question

You can view possible layer names with {external+torchvision:func}`torchvision.models.feature_extraction.get_graph_node_names`. (For more details on the node naming conventions, please see the {external+torchvision:ref}`About Node Names <about-node-names>` heading in the {external+torchvision:doc}`torchvision documentation <feature_extraction>`.)

```python
from torchvision.models import feature_extraction
# this function returns two lists, the first for training mode, the second for eval mode
feature_extraction.get_graph_node_names(deepnet)[1]
```

> And note that you can specify multiple layers!

Reviewer 2: ```suggestion
```

Reviewer 2: I suggested adding this important detail to the main text.

:::

> Next, we grab the preprocessing transform from the model. As the [torchvision docs](https://docs.pytorch.org/vision/stable/models.html#using-the-pre-trained-models) explain it (quoting version `0.27`):

Reviewer 2: ```suggestion
Finally, it is important to specify the preprocessing transform of the model. During neural network training images are transformed before being passed to the network, and the same transformation needs to be applied when using the trained network. As the [torchvision docs](https://docs.pytorch.org/vision/stable/models.html#using-the-pre-trained-models) explain it (quoting version `0.27`):
```

> Before using the pre-trained models, one must preprocess the image (resize with right resolution/interpolation, apply inference transforms, rescale the values etc). There is no standard way to do this as it depends on how a given model was trained. It can vary across model families, variants or even weight versions. Using the correct preprocessing method is critical and failing to do so may lead to decreased accuracy or incorrect outputs.

> For models trained on ImageNet, this preprocessing consists of two steps: resizing to a height and width of 224 pixels and normalizing the color channels (subtracting means and dividing by standard deviations). Following {cite:alp}`Feather2023-model-metam`, we recommend including the normalization step in the model for metamer synthesis, but handling the image resizing externally. We demonstrate how to do so below.

Reviewer 2: ```suggestion
For models trained on ImageNet, this preprocessing consists of two steps: resizing to a height and width of 224 pixels and normalizing the color channels (subtracting means and dividing by standard deviations). Following {cite:alp}`Feather2023-model-metam`, we recommend including the normalization step in the model for metamer synthesis, but handling the image resizing externally.

Both torchvision and timm provide the additional model-specific objects implementing the preprocessing transforms. We show below how to use them to create a normalizing transform to pass to {class}`~plenoptic.models.FeatureExtractorModel`.
```

> Let's grab the normalizing transform and then initialize our plenoptic model:

Reviewer 2: ```suggestion
```

::::{tab-set}
:::{tab-item} torchvision
:sync: torchvision

> In torchvision, the transform is a single torch Module which we cannot easily subdivide, so we create a separate normalization transform, which we pass to {class}`~plenoptic.models.FeatureExtractorModel`:

Reviewer 2: ```suggestion
In torchvision, the transform is a single torch Module which we cannot subdivide, that includes resizing, cropping and normalization. We extract the normalization parameters from this torch Module to create a separate standalone normalization transform:
```

Reviewer 2: I think it was not clear what things the torch Module includes, that we cannot subdivide.

```{literalinclude} ../../scripts/feature_extractor.py
:dedent:
:start-at: weights.transform
:end-at: transforms.Normalize
```

```python
print(transform)
```

```python
ImageClassification(
    crop_size=[224]
    resize_size=[256]
    mean=[0.485, 0.456, 0.406]
    std=[0.229, 0.224, 0.225]
    interpolation=InterpolationMode.BILINEAR
)

> ```

Reviewer 2: Is there a way to make this render more like a printed output? I was very confused because it looked like a class definition.

:::

:::{tab-item} timm
:sync: timm

In timm, the transform can be indexed into, so we can explicitly grab the normalization:

```{literalinclude} ../../scripts/feature_extractor.py
:dedent:
:start-at: = create_transform
:end-at: norm = transform
```

```python
print(transform)
```

```python
Compose(
    Resize(size=256, interpolation=bilinear, max_size=None, antialias=True)
    CenterCrop(size=(224, 224))
    MaybeToTensor()
    Normalize(mean=tensor([0.4850, ...]), std=tensor([0.2290, ...]))

> )

Reviewer 2: Is there a way to make this render more like a printed output? I was very confused because it looked like a class definition.

```

:::
::::

> Finally, we'll pass our neural network, target layer, and preprocessing transform to plenoptic's {class}`~plenoptic.models.FeatureExtractorModel`:

Reviewer 2: ```suggestion
Now we can finally create our model by passing the neural network, target layer, and preprocessing transform to plenoptic's {class}`~plenoptic.models.FeatureExtractorModel`:
```

Reviewer 2: Just changed because I used "Finally" for the step of specifying the normalization transform above.

model = po.models.FeatureExtractorModel(deepnet, target_layer, norm)
img = po.process.center_crop(img, transform.crop_size[0])
```

Let's visualize our resulting image:

```{code-cell} ipython3
po.plot.imshow(img, as_rgb=True);
```

ResNet50 is trained to classify images into one of [1000 categories](https://deeplearning.cms.waikato.ac.nz/user-guide/class-maps/IMAGENET/). Any metamer of an intermediate layer should preserve this classification, which is the output of the final layer; this is one of the criteria that {cite:alp}`Feather2023-model-metam` check for synthesis success. Let's examine that classification now, creating a little helper function:

::::{tab-set}
:::{tab-item} torchvision
:sync: torchvision

```{literalinclude} ../../scripts/feature_extractor.py
:dedent:
:start-at: weights.meta
:end-at: weights.meta
```

:::

:::{tab-item} timm
:sync: timm

```{literalinclude} ../../scripts/feature_extractor.py
:dedent:
:start-at: import urllib
:end-at: read().decode()
```
:::
::::

```{literalinclude} ../../scripts/feature_extractor.py
:dedent:
:start-at: def get_category
:end-at: get_category(img
```

```python
guenon
```

> The category, [guenon](https://en.wikipedia.org/wiki/Guenon), is an Old World monkey. Though it isn't the actual species of the monkey in question (a [Celebes crested macaque](https://en.wikipedia.org/wiki/Celebes_crested_macaque)), it's a reasonable category for it.

Reviewer 2: I suggest some reorganization to impose fewer sidetracks on the reader.

I would move this setup of the classification part to the end, right before where you actually show the classification results.

Additionally, I would move the section "Synthesizing the metamer" before "Understanding the model", so the reader gets the actual main point of the tutorial earlier. I see that the tutorial would show the layer visualization outputs in `po.plot.synthesis_status()` before they are actually explained, but I still think it's best.

In the the section "Understanding the model" I would add two subsections, "Visualizing the features" and something like "Metamer classification".

Author: I agree on everything except for the relative location of the "Synthesizing" and "Understanding" sections. hopefully some of the other changes I've made have helped streamline this, let me know what you think

Reviewer 2: I think it is much better overall after the edits.

I personally would still prefer "Synthesis" first, to get the user to the use case earlier. But I don't feel strongly about this, and the new version is good as is too.

```{code-cell} ipython3
rep = model(img)
print(rep)
print(rep.shape)
```

We have flattened the model representation of the given layer (to support representations from multiple layers simultaneously). If you would like to retrieve the original shape, you can use the {func}`~plenoptic.models.FeatureExtractorModel.convert_to_dict` method:

```{code-cell} ipython3
rep = model.convert_to_dict(rep)
print(rep.keys())
print(rep[target_layer].shape)
```

{class}`~plenoptic.models.FeatureExtractorModel` also has a {func}`~plenoptic.models.FeatureExtractorModel.plot_representation` method, which creates two subplots. The first plots the average across channel, the average spatial representation, while the second averages across space to get a per-channel average representation:

```{code-cell} ipython3
fig, _ = model.plot_representation(rep)
```

## Synthesizing the metamer

:::{warning}
We do not perform synthesis in the exact same way as {cite:alp}`Feather2023-model-metam`. However, the resulting metamer is qualitatively similar. We note the differences below.
:::

> Let us initialize our metamer object using the above image and model. Unlike in {cite:alp}`Feather2023-model-metam`, we are using the mean-squared error (the default for {class}`~plenoptic.Metamer`) as our loss function. We also initialize with a sample of uniformly-distributed noise whose values range from 0 to 1, whereas the paper initialized with "a sample from a normal distribution with a standard deviation of 0.05 and a mean of 0.5". Like that paper, we find better synthesis results if we use a learning-rate scheduler to halve the optimizer's learning rate regularly, using {class}`~torch.optim.lr_scheduler.StepLR` (see the following dropdown for more details):

Reviewer: So what loss function did Feather et al., 2023 use? We should specify it.

Reviewer 2: Agree.

```{code-cell} ipython3
met = po.Metamer(img, model)
met.to(DEVICE)

> met.load(

Reviewer: I’m getting an error after this code block: /home/jenkins/agent/workspace/CCN_neurorse_plenoptic_PR-460/lib/python3.12/site-packages/plenoptic/_synthesize/synthesis.py:562: UserWarning: You will need to call setup() to instantiate scheduler
warnings.warn(

Author: this is a warning, not an error. did it show up as an error? how were you running this?

I'll add a bit of explanation about the warning

Author: actually, I'll suppress the warning

    "gamma": 0.5
}
lr = 3e-2 if target_layer == "layer4" else 1e-2
met.setup(
    optimizer_kwargs={"lr": lr, "amsgrad": False},
    scheduler=scheduler,
    scheduler_kwargs=scheduler_kwargs
)
# by setting stop_iters_to_check=max_iter, we ensure it keeps going through
# all 12k iterations
met.synthesize(max_iter=12000, stop_iters_to_check=12000)
```
:::

```{literalinclude} ../../scripts/feature_extractor.py
:dedent:
:start-at: plot.synthesis_status
:end-at: plot.synthesis_status
```

:::{attention}

Depending upon how zoomed in your browser is, there may be some aliasing artifacts in the appearance of the metamers. If you see faint grid lines, you are encouraged to click on the `png` button to view the figure in its own tab and zoom in to avoid aliasing.

:::

::::{tab-set}
:::{tab-item} layer2
:sync: layer2

```{eval-rst}
.. plot:: scripts/feature_extractor_mpl.py layer2
  :include-source: false
```
:::
:::{tab-item} layer3
:sync: layer3

```{eval-rst}
.. plot:: scripts/feature_extractor_mpl.py layer3
  :include-source: false
```
:::
:::{tab-item} layer4
:sync: layer4

```{eval-rst}
.. plot:: scripts/feature_extractor_mpl.py layer4
  :include-source: false
```
:::
::::

> 

Reviewer: For the representation error over space (and model.plot_representation over space before), it could be beneficial to add a small colorbar showing the range and corresponding values. The value range in the title is pretty hard to read.

Author: none of the other similar plots have error bars, so I'm not going to here. I'll add an example in the doctest showing how users can do it after.

Reviewer: Thanks for adding them. Now you've added them I feel like it's a bit extra. Up to you whether you want to keep it. I don't have strong opinions on this, especially if it's better to be consistent with everything else in the documentation.

In the above plots, we can see the metamer in the leftmost subplot, the loss over synthesis iterations in the middle, and the representation error on the right:
- Our metamers match the results discussed earlier in this notebook:  the layer 2 metamer looks almost identical to the target image, the layer 3 metamer starts to add RGB noise, and the layer 4 is almost completely unidentifiable, looking almost completely like random RGB noise.

> - We can see that the optimization performed reasonably well: the loss decreased gradually over synthesis. If you were using these stimuli in an experiment (especially for `"layer4"`), it may be worth continuing a bit more to get the loss even lower, but these demonstrate the point.

Reviewer: It is also interesting the converged loss increases as you go to deeper layers. What’s the intuition behind this? Did the paper show this as well?

Author: I do not have an intuition here. The paper doesn't discuss the absolute loss value, or if any of their success metrics differ across layers.

The paper uses a relative loss (dividing the Euclidean distance by the L2-norm of the representation from the layer), and it's possible that doing that would normalize this value.

> - The representation error plot has the same structure as the {func}`~plenoptic.models.FeatureExtractorModel.plot_representation` plot above. We see that the error is fairly uniform across both space and channels.

Reviewer: I also wouldn’t say that error is uniform.. there is clearly hotspots in space.

Author: Good point, I've updated this wording

The authors of {cite:alp}`Feather2023-model-metam` used two additional checks to verify that metamer synthesis had succeeded (quotes from "Results > Metamer optimization" section, pdf page 5):
- "the metamer had to result in the same classification decision by the model as the reference stimulus" (here, `guenon`):

> - "measures of the match between the activations for the natural reference stimulus and its model metamer at the matched stage had to be much higher than would be expected by chance, as quantified with a null distribution". The authors used three measures here: Pearson and Spearman correlations and signal-to-noise ratio. Here, we show the Pearson correlation:

Reviewer: Like you mentioned, the correlation values on their own can’t verify success unless compared against the distribution of correlations from the null distribution. Would that be an easy addition?

Author: unfortunately, no. That would require getting all the images they used. I've tweaked the wording a bit
